SIP-030: Wallet RPC Standards

janniks opened the pull request:
## Wallet RPC Standards
📓 [Read the full SIP-030](https://github.com/janniks/sips/blob/main/sips/sip-030/sip-030-wallet-interface.md) on Wallet RPC Standards

- Adds SIP for more modern standard for wallet API RPC (usable for browser-extensions as well as other settings)
- Proposes standards for serialization of wallet RPC requests

---

> Currently being reviewed and edited for final approval

janniks committed:
feat: add wallet api sip

janniks commented on sips/sip-02X/sip-02X-wallet-interface.md:
```diff
@@ -0,0 +1,213 @@
+- `attachment?`: `string` hex-encoded
+- `anchoreMode?`: `'onChainOnly' | 'offChainOnly' | 'any'`
+- `postConditionMode?`: `'allow' | 'deny'`
+- `postConditions?`: `PostCondition[]`, defaults to `[]`
```
I have tried to align with the JSON API from Xverse, but in some cases I believe it's likely better to decouple from Stacks.js representations of data (post-conditions, clarity-values, etc.)
It should easily be possible for Xverse to add handling for a backward compatible interface. cc @m-aboelenein

janniks replied:
@m-aboelenein do you think that's something Xverse could support -- I think it's important to stay disconnected from the JavaScript representations _(since those might not exist everywhere or could change)_

aryzing replied:
From the "Notes on Serialization" section, `postConditions` would be typed as a `string[]`?

aryzing replied:
Just saw the "where" section below regarding encoding PostCondition as hex-encoded

janniks replied:
Yeah, still discussing these. Not sure what's best, but we're thinking for Clarity hex or basic JS objects (similar to Stacks.js, but with readable enums) could work. For post-conditions, I'm not sure if there is a clean solution, the JS repr isn't really nice, but could also work similarly to Clarity.

janniks committed:
fix: add note on namespaces

kyranjamie commented on sips/sip-02X/sip-02X-wallet-interface.md:
```diff
@@ -0,0 +1,216 @@
+
+`result`
+
+- `txId`: `string` hex-encoded
```
IMO in Bitcoin the more common treatment of this is all lower `txid`, as it's treated as a single concept or word https://github.com/search?q=repo%3Abitcoin%2Fbitcoin%20txid&type=code

janniks replied:
I agree. `txid` is nice, should we also do `txRaw`/`txHex` → `tx` (or `transaction`)?

kyranjamie replied:
I think that makes makes yeah. Typings make describing the format largely unnecessary.

janniks committed:
fix: update tx results

kyranjamie commented:
Great work @janniks. Leather supports this SIP 🖤

janniks committed:
chore: minor updates

m-aboelenein commented on sips/sip-02X/sip-02X-wallet-interface.md:
```diff
@@ -0,0 +1,213 @@
+- `postConditionMode?`: `'allow' | 'deny'`
+- `postConditions?`: `PostCondition[]`, defaults to `[]`
+- `sponsored?`: `boolean`, defaults to `false`
+- `sender?`: `string` address, Stacks c32-encoded
```
changing it to address is fine

janniks replied:
👍 updated thx

janniks committed:
chore: add open questions

janniks commented:
Thanks for the input @m-aboelenein @kyranjamie .
I updated the stuff we chatted about on the call, and also added some `OPEN QUESTIONS` to the top.

Do you think you could do another review with final remarks until ~Tue. 20.02.2024 to wrap this up? 🙏

janniks committed:
fix: update address and account methods

janniks commented:
Added methods `getAddresses` and `getAccounts` for better compatibility with current solution.

janniks committed:
chore: update sip

aryzing commented on sips/sip-02X/sip-02X-wallet-interface.md:
```diff
@@ -0,0 +1,244 @@
+Bytes are serialized as hex-encoded strings (without a 0x prefix).
+Predefined formats from previous SIPs are used where applicable.
+Addresses are serialized as Stacks c32-encoded strings.
+Clarity values, post-conditions, and transactions are serialized to bytes (defined by SIP-005) and used as hex-encoded strings.
```
What about post-condition mode?

aryzing commented on sips/sip-02X/sip-02X-wallet-interface.md:
```diff
@@ -0,0 +1,244 @@
+`params`
+
+- `message`: `string` Clarity value, hex-encoded
+- `domain?`: `string` hex-encoded (defined by SIP-018)
```
What's making the `domain` be optional? Shouldn't `stx_signMessage` be used when there's no domain? Alternatively, would it make sense to have a single `stx_signMessage` method with an optional `domain?` param?

aryzing commented on sips/sip-02X/sip-02X-wallet-interface.md:
```diff
@@ -0,0 +1,244 @@
+- `message`: `string` Clarity value, hex-encoded
+- `domain?`: `string` hex-encoded (defined by SIP-018)
+
+> `domain` can be optional if the wallet (e.g. browser extension) can infer it from the origin of the request.
```
In the context of SIP-18, doesn't "domain" refer to something else, not the request's origin?

Domain should be a tuple no?

```clarity
{
	name: (string-ascii len)
	version: (string-ascii len)
	chain-id: uint
}
```

Wouldn't auto-populating a domain mean wallets would be signing messages with data users did not ask to be included?

aryzing commented on sips/sip-02X/sip-02X-wallet-interface.md:
```diff
@@ -0,0 +1,244 @@
+- `signature`: `string` hex-encoded
+- `publicKey`: `string` hex-encoded
+
+### `stx_signStructuredMessage`
```
Should we include a `result` section? Would it be the same as the `result` section for `stx_signMessage`?

janniks replied:
Yep, will do -- same same

aryzing commented on sips/sip-02X/sip-02X-wallet-interface.md:
```diff
@@ -0,0 +1,244 @@
+
+- `contractName`: `string`
+- `codeBody`: `string` Clarity contract code
+- `clarityVersion?`: `number`
```
Would it be better to have it as a string to allow for semver or similar?

janniks replied:
Right now it's encoded as an integer (on the contract deploy side), so number should be the JS closest representation.

janniks committed:
docs: add new js representation

janniks commented:
@aryzing @kyranjamie I added an [update](https://github.com/stacksgov/sips/pull/166/commits/425f81ca2aacab06f92c84b8a88023a94a2a6b82) to the encoding (as discussed on last weeks call).
Let me know what you think! This can be used as the canonical repr and Stacks.js would also migrate towards it. (Serialized hex encoded string could also still be used (easy to tell apart).

Also updated:
- renamed `functionName` to `name`, `functionArgs` to `arguments`
- merged `contractName` + `contractAddress` to `contract` -- which is a `${string}.${string}` (how it's copy pasted from the explorer)
- renamed some other params (see 425f81ca2aacab06f92c84b8a88023a94a2a6b82 for details)

Ping me if you disagree with anything -- trying to make the naming more logical and short

pradel commented:
@janniks I like this change, less magic on what's going on and the behavior seems to be pretty similar to how ETH is doing things, making it easier for ETH devs to use the API

aryzing commented on sips/sip-02X/sip-02X-wallet-interface.md:
```diff
@@ -218,7 +221,232 @@ The following definitions can be used in multiple methods (mainly for transfer a
 - `profile`: `object` updated Schema.org Person object
 
-## Provider Registration
+## JS Representations
```
Will definitely provide a better dev experience by making the values easier to inspect and transmit. Looking forward to using these JS representations.

janniks commented on sips/sip-02X/sip-02X-wallet-interface.md:
```diff
@@ -0,0 +1,472 @@
+`params`
+
+- `contract`: `string.string` address with contract name suffix, Stacks c32-encoded
+- `function`: `string`
```
Is `arguments` reserved? I had this feeling as well, thought it might not be great, but wanted something shorter ☹️

Fine with going back to functionName, functionArgs. But happy to consider alternatives

janniks replied:
Changed to `functionName` and `functionArgs` again.
We could do something with `public` (since only public functions can be called. Any ideas welcom

janniks committed:
docs: update methods

janniks committed:
docs: update open questions

janniks commented:
Also, Thanks for all the feedback everyone! 👏

I feel like we're nearing a good document here. I think we can wrap it up soon 🫡

kyranjamie commented:
@janniks sips@kyranjamie.com & @kyranjamie

aryzing commented:
> Let me know which email-addresses/github-usernames to add to the author list

Aren't all the commits yours? :joy: I wasn't even thinking about making it to the author list, not sure I've contributed enough, up to you :shrug:

janniks commented:
I’ll consider anybody an author/editor that contributed in some say 😉 so lmk. Honestly I don’t need the section, but probably even changeable later

aryzing commented:
Sure, why not :) link to my github, @aryzing , thanks :pray:

314159265359879 commented:
@janniks does this modern wallet API, include API's that would help the wallet/dapp know if there is a difference between the (A) network and/or (B) account between what is active on the extension and what is signed in with on the dapp?
Say you are on account 1 in the dapp and your Leather is active on account 2. In Metamask you would get a warning (do you want to switch to account 2?) or something like that. I have a feeling that could help prevent some issues where users are confused about which account is signing. 

A. Account switching, There may be other relevant issues, but this is a recent one that came to my mind https://github.com/leather-wallet/extension/issues/4859
As part of that, this one is probably harder/impossible? (detect when different secret key is used) https://github.com/leather-wallet/extension/issues/3006

Similar with different networks being used.
Especially relevant when users are on testnet and forget to switch back to mainnet... the experience is currently confusing for users. This issue pops up every time there is a popular public testnet. 

I think we'll see similar issues once subnets and other layered approaches become popular which require users to install/add networks much like Metamask does for L2 on Ethereum.
B. Network: https://github.com/leather-wallet/extension/issues/4194 (changing network, adding network, dapp to prompt switch or addition)

m-aboelenein commented:
@janniks @m-aboelenein & mahmoud@xverse.app

m-aboelenein commented:
Exciting update! I am thrilled to share that we now have a beta version of sats-connect that implements the SIP check the docs for details. https://docs.xverse.app/sats-connect-wallet-api-for-bitcoin-and-stacks-1
Very excited to hear your thoughts and feedback 🙌

aryzing commented on sips/sip-02X/sip-02X-wallet-interface.md:
```diff
@@ -0,0 +1,500 @@
+
+`result`
+
+- `txid`: `string` hex-encoded
```
Seems that many of the param names in this spec are camelCased, so doesn't `txid` stand out as not following this general pattern? If we're already using a `transaction` parameter in many calls, why not use `transactionId` here? It would keep in line with the camelCasing, as well as JS's generally verbose names. 

From what I've seen, codebases dealing with transactions use both `transaction` and `tx` in variable names (and so does this spec!), so I'd propose sticking with `transaction`, and help maintain the use of `transaction` in various related specs by updating the [WBIP003 draft](https://webbtc.netlify.app/wbips/WBIP003) to include a case for the term `transaction`. Wdyt?

janniks replied:
We had this discussion a while back, it was `txId` before, but since txid is a term on its own in the bitcoin/blockchain world, we went with `txid`

aryzing replied:
I guess where I was coming from with the suggestion is that programming environments do their best to refer to the same ideas using their own conventions. For a transaction ID, I'd assume that `txid` would be more in line with how variables are named in lower level languages. And so in this spec it kind of stands out for not using camelCase as is more typical of JS environments, plus we're already using the word `transaction` in other areas :sweat_smile:

janniks replied:
I get that. Both make sense to me. `txid` is a bit more Bitcoin aligned. It's referred to as a word/concept/thing in the core node etc. https://github.com/search?q=repo%3Astacks-network%2Fstacks-core%20txid&type=code so I feel it's what developers expect.

aryzing commented:
> does this modern wallet API, include API's that would help the wallet/dapp know if there is a difference between the network and/or account between what is active on the extension and what is signed in with on the dapp?

Doesn't this imply that the account management strategy / state management strategy of wallets would leak into the spec? Seems that the concept of an "active" account is just a way for some wallets to organize or handle what data they keep in memory or present to the user. A "comprehensive" wallet could present all accounts & balances simultaneously to the user (even from both mainnet & testnet). 

Would it make sense to include the address & network as part of the request? After all, if the result of the operations in this SIP are a function of the network and address, should those not be included in the requests too?

If included, the "comprehensive" wallet described above (that has no concept of an "active" account) would know exactly what to do too. For wallets segregating state based on the "active" account abstraction, they'd have a clear way of checking whether address / network match the currently "active" ones and act as seen fit.

janniks commented:
There was similar ideas to https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1193.md#chainchanged-1 -- we haven't scoped them onto this SIP yet, but we can add it.

janniks committed:
fix: typo

Co-authored-by: Marvin <MarvinJanssen@users.noreply.github.com>

janniks committed:
fix: typo

Co-authored-by: Marvin <MarvinJanssen@users.noreply.github.com>

janniks committed:
docs: update sip

janniks committed:
docs: update sip

janniks committed:
chore: add comment

janniks committed:
chore: address pr comments

janniks committed:
chore: update authors

janniks committed:
chore: update

janniks commented on sips/sip-02X/sip-02X-wallet-interface.md:
```diff
@@ -0,0 +1,500 @@
+- `txid`: `string` hex-encoded
+- `transaction`: `string` hex-encoded raw transaction
+
+### Method `stx_transferFt`
+
+`params`
+
+- `recipient`: `string` address, Stacks c32-encoded
+- `asset`: `string` address, Stacks c32-encoded, with contract name suffix
+- `amount`: `number | string` BigInt constructor compatible value
+
+`result`
+
+- `txid`: `string` hex-encoded
+- `transaction`: `string` hex-encoded raw transaction
```
Yes, good point. Maybe better to leave these out for now. 
Or rename to `stx_transferSip10Ft`, `stx_transferSip10Nft` 🤔

314159265359879 commented:
> There was similar ideas to https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1193.md#chainchanged-1 -- we haven't scoped them onto this SIP yet, but we can add it.

address and network as part of the request, I would love to see that scoped in. @aryzing thanks for that explanation I do believe that is a valuable addition to this sip to help deal with the issues I mentioned with the wallet.

janniks committed:
chore: add wbip links

janniks committed:
chore: add open question

kyranjamie commented on sips/sip-02X/sip-02X-wallet-interface.md:
```diff
@@ -0,0 +1,527 @@
+- `provider.listen(event: string, listener: (...args: any[]) => void): void`
+- `provider.unlisten(event: string, listener: (...args: any[]) => void): void`
+
+### Event `accountsChanged`
```
I think we should avoid `-ed` past tense naming. Browser for instance uses `click` events not `onclicked`.

`onNetworkChange`, `accountChanged`?

(Unsure about `on` prefix)

aryzing replied:
I think the events names are closer to nouns than verbs, e.g.,

- A `click` happened
- A `blur` happened
- An `accountChange` happened

hence no -ed. 

I see the `on` prefix used for the event handlers, e.g., `onmousedown="..."`, not for event names. [Some HTML spec event names](https://html.spec.whatwg.org/multipage/webappapis.html#handler-onclick) for inspiration :)

janniks committed:
fix: pr comment changes

kyranjamie commented on sips/sip-02X/sip-02X-wallet-interface.md:
```diff
@@ -123,7 +122,7 @@ The following definitions can be used in multiple methods (mainly for transfer a
 - `transaction`: `string` hex-encoded raw transaction
 
-### Method `stx_transferFt`
+### Method `stx_transferSip10Ft`
```
👍🏼

janniks committed:
fix: add origin post conditions

Hero-Gamer commented:
Helping SIP Editors with SIP number assignment.
SIP will be assigned to: **SIP-030** @janniks 

SIP Editors @314159265359879 @rafaelcr @jiga @AcrossfireX please help review and once ready request "Accepted" status.

janniks committed:
chore: rename to 30

wileyj commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,528 @@
+
+Title: Integration of a Modern Stacks Wallet Interface Standard
+
+Authors: [janniks](https://github.com/janniks), [kyranjamie](https://github.com/kyranjamie), [aryzing](https://github.com/aryzing), [m-aboelenein](https://github.com/m-aboelenein)
```
Sort alphabetically please

wileyj commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,528 @@
+
+# Abstract
+
+This SIP proposes the integration of the standard into the Stacks blockchain's "Connect" and "Auth" system.
```
This is a bit confusing: "...integration of the standard into the Stacks..."
in this context, what is *the standard*?

wileyj commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,528 @@
+# Abstract
+
+This SIP proposes the integration of the standard into the Stacks blockchain's "Connect" and "Auth" system.
+The goal is to replace the current Connect interface, primarily used by web applications to connect with browser extensions and mobile apps, with a more straightforward protocol.
```
```suggestion
The goal is to replace the current Connect interface, primarily used by web applications to connect with browser extensions and mobile apps with a more straightforward protocol.
```

wileyj commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,528 @@
+
+This SIP proposes the integration of the standard into the Stacks blockchain's "Connect" and "Auth" system.
+The goal is to replace the current Connect interface, primarily used by web applications to connect with browser extensions and mobile apps, with a more straightforward protocol.
+The proposal consists mainly of standardizing JSON compatible interfaces for use with wallet interfaces.
```
```suggestion
This proposal's goal is to standardize JSON compatible interfaces for use with wallet interfaces.
```

wileyj commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,528 @@
+
+# Motivation
+
+The current Connect system, which has evolved since the Blockstack era, is primarily utilized by web applications for interfacing with wallets.
```
Unless this will be annotated or linked, i would change "which has evolved since the Blockstack era" to something more generic like: "which has existed for several years, ..."
or, remove that part of the sentence altogether unless it's more relevant later on (still reading)

wileyj commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,528 @@
+# Motivation
+
+The current Connect system, which has evolved since the Blockstack era, is primarily utilized by web applications for interfacing with wallets.
+However, many aspects of the existing "Connect" and "Auth" libraries are no longer required, leading to unnecessary complexity and brittleness in wallet connectivity.
```
Can you list a few things that are no longer required to give the reader of how this may be brittle?

wileyj commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,528 @@
+The current Connect system, which has evolved since the Blockstack era, is primarily utilized by web applications for interfacing with wallets.
+However, many aspects of the existing "Connect" and "Auth" libraries are no longer required, leading to unnecessary complexity and brittleness in wallet connectivity.
+
+Recent attempts to standardize the interface have sparked valuable discussions but have not culminated in a ratified standard, largely due to the stable state of the existing system.
```
Assuming there are public discussions here, can you annotate and link to them?

wileyj commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,528 @@
+This SIP aims to address these issues by adopting the WBIPs standards, which offer a more suitable RPC-style interface for modern web applications.
+The simplified protocol will allow integration without heavy dependencies (like Auth) and provider a more extendable interface for wallets.
+
+Additionally, this SIP is motivated by the increased traffic of Ordinal inscriptions on Bitcoin and the Stacks ecosystem growing closers to Bitcoin.
```
```suggestion
Additionally, this SIP is motivated by the increased traffic of Ordinal inscriptions on Bitcoin and the Stacks ecosystem growing closer to Bitcoin.
```

wileyj commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,528 @@
+
+Additionally, this SIP is motivated by the increased traffic of Ordinal inscriptions on Bitcoin and the Stacks ecosystem growing closers to Bitcoin.
+The community has recognized the need for a more unified approach to wallet connectivity (e.g. Bitcoin and PSBTs for previously Stacks-only wallets).
+By adopting the new standard, we aim to align the community towards a common and modern protocol for wallet interaction in web applications.
```
this is interesting - does this proposal only affect these web application interactions, or would offline users also benefit from this change?

janniks replied:
Theoretically, the RPC standard can be used anywhere. e.g. we could have the CLI options align with the RPC as well. But web is the largest driver (although the sip intends to be more or less runtime agnostic)

wileyj commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,528 @@
+
+The proposed changes are listed as follows:
+
+Specify JSON-RPC 2.0 compatible methods and payloads for wallet interaction.
```
nit: please add a link or annotation to the JSON-RPC 2.0 spec

wileyj commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,528 @@
+# Backwards Compatibility
+
+The implementation of this proposal is not necessarily backward compatible.
+However, wallets implementing the new standard are advised to maintain the previous system to support legacy applications during a transition period.
```
how will they be advised here?
maybe reword slightly:
```
...new standard should also maintain the previous...
```

if we leave in "are advised", it's something that should be mentioned in the SIP as to how that advisory will take place.

wileyj commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,528 @@
+
+The implementation of this proposal is not necessarily backward compatible.
+However, wallets implementing the new standard are advised to maintain the previous system to support legacy applications during a transition period.
+Existing applications using the current Auth system should continue to operate, but immediate changes are recommended once this SIP is ratified.
```
"should" or "shall"?
`should` implies we don't know if it would work. i think this is a wording issue, i can't image a wallet would stop working immediately if this proposal is approved.

wileyj commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,528 @@
+The implementation of this proposal is not necessarily backward compatible.
+However, wallets implementing the new standard are advised to maintain the previous system to support legacy applications during a transition period.
+Existing applications using the current Auth system should continue to operate, but immediate changes are recommended once this SIP is ratified.
+The Connect library should go live with the new standard, once most major wallets have adopted the new standard.
```
I don't like the wording here, it opens up a lot of potential questions that aren't likely to be in this SIP. 

i think you can reword it to reflect a timeframe, rather than a nebulous "once most major wallets have adopted". 
otherwise, i think you'll have to enumerate which wallets will need to be updated, and how much time will be provided for that upgrade.

janniks replied:
True, I'll try to add it to the ratifying section

wileyj commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,528 @@
+
+## Notes on Serialization
+
+These methods are based on JSON-RPC 2.0 for serialization.
```
I don't think this is needed - JSON-RPC is already part of the specification section. it's implied that serialization would follow that standard.

wileyj commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,528 @@
+## Notes on Serialization
+
+These methods are based on JSON-RPC 2.0 for serialization.
+To adhere to a generic serializability, the following notes are given.
+Enums are serialized as human-readable strings.
+BigInts are serialized as numbers, strings, or anything that can be parsed by the JavaScript BigInt constructor.
+Bytes are serialized as hex-encoded strings (without a 0x prefix).
+Predefined formats from previous SIPs are used where applicable.
+Addresses are serialized as Stacks c32-encoded strings.
+Clarity values, post-conditions, and transactions are serialized to bytes (defined by SIP-005) and used as hex-encoded strings.
```
```suggestion
To adhere to a generic serializability, the following notes are given:
- Enums are serialized as human-readable strings.
- BigInts are serialized as numbers, strings, or anything that can be parsed by the JavaScript BigInt constructor.
- Bytes are serialized as hex-encoded strings (without a 0x prefix).
- Predefined formats from previous SIPs are used where applicable.
- Addresses are serialized as Stacks c32-encoded strings.
- Clarity values, post-conditions, and transactions are serialized to bytes (defined by SIP-005) and used as hex-encoded strings.
```

wileyj commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,528 @@
+This section defines the available methods, their parameters, and result structure.
+Parameters should be considered recommendations for the wallet.
+The user/wallet may choose to ignore/override them.
+Optional params are marked with a `?`.
```
```suggestion
*Note: *Optional params are marked with a `?`.
```

wileyj commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,528 @@
+Optional params are marked with a `?`.
+
+Methods can be namespaced under `stx_` if used in more generic settings and other more Ethereum inspired domains.
+In other cases (e.g. WalletConnect), the namespace may already be given by meta-data (e.g. a `chainId` field) and can be omitted.
```
```suggestion
In other cases (e.g. `WalletConnect`), the namespace may already be given by meta-data (e.g. a `chainId` field) and can be omitted.
```

wileyj commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,528 @@
+
+---
+
+### Method `stx_transferStx`
```
The structure here seems a little ill-suited. have you considered a table vs markdown lists?
the text for all of the methods seems fine, i just think the data can be presented to the reader in an easier to consume format.

wileyj replied:
looking it over again, maybe a table is also ill-suited. 
perhaps some better formatting though?
was thinking about this, ex:
```
### stx_transferSip10Ft

`params`

- `recipient`: `string` address, Stacks c32-encoded
- `asset`: `string` address, Stacks c32-encoded, with contract name suffix
- `amount`: `number | string` BigInt constructor compatible value

`result`

- `txid`: `string` hex-encoded
- `transaction`: `string` hex-encoded raw transaction
```

janniks replied:
Yeah, table is also tricky. I'll try alternate headings and change how the `where` adds an additional unneeded section.

wileyj commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,528 @@
+
+> **Comment**: For encoding larger than JS `Number` big integers, `string` is used.
+
+`0x00` `int`
```
nit: this are of type json, not ts

wileyj commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,528 @@
+
+## Provider registration
+
+Wallets can register their aliased provider objects according to WBIP-004.
```
annotate of provide link to this resource please

wileyj commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,528 @@
+
+# Ratification
+
+This SIP is considered ratified after at least two major wallets in the Stacks ecosystem have implemented and launched the new standard.
```
How shall this be independently verified? I'm leaning more towards a timeframe rather than a self-reported metric here.

janniks replied:
This part I'm not sure about, yeah. How is it typically done for SIPs?

> Note: Even if it's not widely used yet, wallets are already implementing a lot of this in collaboration of the SIP.

wileyj commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,528 @@
+
+Wallets can register their aliased provider objects according to WBIP-004.
+
+# Ratification
```
formatting - all of these H1 markdown headers should be H2 (h2->h3 and so on)

wileyj commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,528 @@
+
+This SIP is considered ratified after at least two major wallets in the Stacks ecosystem have implemented and launched the new standard.
+
+# Links
```
```suggestion
## Appendix
```

wileyj replied:
Additionally, i would annotate this section and link where they're referenced above

wileyj commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,528 @@
+By adopting the new standard, we aim to align the community towards a common and modern protocol for wallet interaction in web applications.
+Importantly, the decision to use an existing standard (rather than designing a new one or reworking Auth) is intentional — to avoid further division or split ownership within the community.
+
+There was an attempt to re-use existing standards/protocols from other ecosystems via the WBIPs working group — but no consensus was found that was a perfect fit or had enough traction for the larger layer-2 ecosystem.
```
If those discussions are public, can you annotate/link here?

wileyj commented on sips/sip-030/sip-030-wallet-interface.md:
````diff
@@ -0,0 +1,528 @@
+```ts
+{
+  type: 'contract',
+  value: `${string}.${string}` // Stacks c32-encoded, with contract name suffix
````
this is in several places, but are backticks intended? in a codeblock, they show up as typed

janniks replied:
Yes, this is a typescript string type, which essentially constrains string type to a subset, matching the expected criteria (not doing much in this case, e.g. it would likely only fail on a string ending with a period; but it's also nice for developer understanding and type hinting in IDEs — this same type could be used in TS codebases)

janniks committed:
chore: update from pr review

janniks committed:
chore: add pr feedback

janniks committed:
chore: add pr feedback

friedger commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,527 @@
+- `network?`: `'mainnet' | 'testnet' | 'regtest' | 'mocknet'`
+- `fee?`: `number | string` BigInt constructor compatible value
+- `nonce?`: `number | string` BigInt constructor compatible value
+- `attachment?`: `string` hex-encoded
```
Are attachments still a thing? We moved on to StackerDBs

jcnelson replied:
I've been considering opening a (separate) SIP to deprecate the Atlas subsystem now that nothing is using it.

In that vein, it may be more appropriate to call this field something non-Atlas-specific, like `content-hash`.  Then, it could be used by clients to authenticate off-chain content obtained from StackerDB or somesuch system.

janniks replied:
If it's not super clear when and how the content hashes would be used going forward, I'll err on the side of removing it completely (for now) — wallets can still decide to add it later

jiga commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,527 @@
+Title: Integration of a Modern Stacks Wallet Interface Standard
+
+Authors: [aryzing](https://github.com/aryzing), [janniks](https://github.com/janniks), [kyranjamie](https://github.com/kyranjamie),  [m-aboelenein](https://github.com/m-aboelenein)
+
```
Missing 'Consideration' field in the Preamble:  Please add this field and specify the appropriate consideration(s), such as 'Technical'.

janniks replied:
For the `Type` field is there a category "non consensus breaking standard"?

jiga commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,527 @@
+This SIP proposes common RPC methods to use for the Stacks blockchain's "Connect" and "Auth" systems.
+The goal is to replace the current Connect interface, primarily used by web applications to connect with browser extensions and mobile apps with a more straightforward protocol.
+This proposal's goal is to standardize JSON compatible interfaces for use with wallet interfaces.
+
```
Missing 'Copyright' section here

janniks replied:
Haven't seen that in other sips, what should it look like?

wileyj replied:
https://github.com/stacksgov/sips/blob/main/sips/sip-000/sip-000-stacks-improvement-proposal-process.md?plain=1#L156

jiga commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,527 @@
+The goal is to replace the current Connect interface, primarily used by web applications to connect with browser extensions and mobile apps with a more straightforward protocol.
+This proposal's goal is to standardize JSON compatible interfaces for use with wallet interfaces.
+
+# Motivation
```
'Motivation' section should be renamed 'Introduction'?: The SIP Specification requires an 'Introduction' section that provides a high-level summary of the problem(s) and how the proposal solves them.

jiga commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,527 @@
+
+Specify [JSON-RPC 2.0](https://www.jsonrpc.org/specification) compatible methods and payloads for wallet interaction.
+These can be used via a browser object (i.e., via the `window.btc.request` method) or similar interfaces like WalletConnect.
+
```
Missing 'Related Work' section: summarize alternative solutions and explain why they are not adequate.

jiga commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,527 @@
+The implementation of this proposal is not necessarily backward compatible.
+Wallets implementing the new standard may maintain the previous system to support legacy applications during a transition period or indefinitely.
+Existing applications using the current Auth system should continue to operate, but immediate changes are recommended once this SIP is ratified.
+
```
Missing 'Activation' section: please describe the timeline, criteria, and process for activating the SIP once it is ratified.

wileyj replied:
@janniks this section is where you can define your existing "ratification" section. and example you may use is: https://github.com/stacksgov/sips/blob/main/sips/sip-012/sip-012-cost-limits-network-upgrade.md#activation-criteria

janniks replied:
Renamed to activation 👍

jiga commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,527 @@
+Wallets implementing the new standard may maintain the previous system to support legacy applications during a transition period or indefinitely.
+Existing applications using the current Auth system should continue to operate, but immediate changes are recommended once this SIP is ratified.
+
+# Implementation
```
rename to 'Reference Implementations' ?

jiga commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,527 @@
+
+# OPEN QUESTIONS
+
+<!-- todo: remove section before merge -->
```
reminder to address these and remove open questions section

jiga commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,527 @@
+
+Recent attempts to standardize the interface (https://github.com/stacksgov/sips/pull/59, https://github.com/stacksgov/sips/issues/117) have sparked valuable discussions but have not culminated in a ratified standard, largely due to the stable state of the existing system.
+This SIP aims to address these issues by adopting the WBIPs standards, which offer a more suitable RPC-style interface for modern web applications.
+The simplified protocol will allow integration without heavy dependencies (like Auth) and provider a more extendable interface for wallets.
```
```suggestion
The simplified protocol will allow integration without heavy dependencies (like Auth) and provide a more extendable interface for wallets.
```
fix typo

jiga commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,527 @@
+
+While discussing this SIP, it has become clear that the current Stacks.js representation is confusing to developers.
+Rather, a better solution would be human-readable — for example, rely on string literal enumeration, rather than magic values, which need additional lookups.
+Relying on soley a hex-encoded also poses difficulties when building Stacks enabled web applications.
```
```suggestion
Relying solely on hex-encoded also poses difficulties when building Stacks-enabled web applications.
```
fix grammar/spelling

jiga reviewed (changes requested):
SIP editor review

kyranjamie commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,527 @@
+The proposed changes are listed as follows:
+
+Specify [JSON-RPC 2.0](https://www.jsonrpc.org/specification) compatible methods and payloads for wallet interaction.
+These can be used via a browser object (i.e., via the `window.btc.request` method) or similar interfaces like WalletConnect.
```
Should we remove reference to `.btc`? 

```suggestion
These can be used via a browser object (i.e., via the `window.WalletProvider.request` method) or similar interfaces like WalletConnect.
```

janniks committed:
chore: update sip

Co-authored-by: kyranjamie <kyranjamie@users.noreply.github.com>

janniks committed:
chore: update sip

Co-authored-by: kyranjamie <kyranjamie@users.noreply.github.com>

janniks committed:
chore: pr feedback

janniks committed:
chore: add more references

janniks committed:
chore: add wallet standard

janniks committed:
chore: pr feedback

janniks committed:
chore: pr feedback

janniks commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,527 @@
+
+Methods can be namespaced under `stx_` if used in more generic settings and other more Ethereum inspired domains.
+In other cases (e.g. `WalletConnect`), the namespace may already be given by metadata (e.g. a `chainId` field) and can be omitted.
+On the predominant `StacksProvider` global object, the methods can be used without a namespace, but wallets may add namespaced aliases for convenience.
```
That's definitely an idea. We should check with Xverse and discuss the latest thoughts on wallet-standard and Reown as well.

janniks committed:
chore: pr feedback

janniks committed:
chore: pr feedback

janniks committed:
chore: pr feedback

janniks committed:
chore: pr feedback

janniks committed:
chore: pr feedback

friedger commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,532 @@
+
+### Provider registration
+
+Wallets can register their aliased provider objects however they see fit.
```
Is this specific enough? Currently, different wallets use different methods and are not working nicely together.

janniks replied:
Yeah, I agree we should add more on provider discovery.

kyranjamie commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,532 @@
+- `txid`: `string` hex-encoded
+- `transaction`: `string` hex-encoded raw transaction
+
+#### Method `stx_transferSip10Nft`
```
Aren't NFTs Sip9?

friedger replied:
yes

kyranjamie commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,532 @@
+
+#### Event `accountChange`
+
+`listener: (accounts: {}[]) => void`
```
Wondering if this is a privacy issue. A user might only wish to share one particular account details with a given app, but this switcher api implicitly shares all account details, (if a user switches to all accounts).

Instead, perhaps we should return only the account index, such that the app can know the account has changed.

janniks committed:
fix: update sip properties

314159265359879 commented:
I liked the previous title of this standard better because it was easier for a lay person to understand. 

What do you think about one of these alternatives instead?
1. Unified Wallet Communication Standards
`(Highlights the goal of creating a single standard for wallet communication without technical jargon.)`
2. A Modern Standard for Stacks Wallet Connections
`(Conveys the update and relevance of the proposal.)`
3. Simplifying Wallet Interactions Across Platforms in the Stacks Network
`(Focuses on the user-friendly impact of the standard.)`

friedger commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,538 @@
+
+## Backwards Compatibility
+
+The implementation of this proposal is not necessarily backward compatible.
```
Can we mention that most importantly, app private keys for domains have been removed.

janniks committed:
fix: fix sip number reference

wileyj commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,538 @@
+## Introduction
+
+The current Connect system[^15], which has existed for several years, is primarily utilized by web applications for interfacing with wallets.
+However, many aspects of the existing "Connect" and "Auth" libraries are no longer required, leading to unnecessary complexity (e.g., wrapping RPC payloads in jsontokens) and lack of clear definitions (e.g., undefined serialization for non-JSON compatible data structures) in wallet connectivity.
```
`Auth` is mentioned several times in this SIP. i think it would be helpful to define these libraries in a related work section (exists later in this sip): https://github.com/stacksgov/sips/blob/main/sips/sip-000/sip-000-stacks-improvement-proposal-process.md?plain=1#L171

wileyj commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,538 @@
+This SIP aims to address these issues by adopting the WBIPs standards[^11], which offer a more suitable RPC-style interface for modern web applications.
+The simplified protocol will allow integration without heavy dependencies (like Auth) and provide a more extendable interface for wallets.
+
+Additionally, this SIP is motivated by the increased traffic of Ordinal inscriptions on Bitcoin and the Stacks ecosystem growing closer to Bitcoin.
```
actually, i think the next sentence addresses my concern, but i would slightly change it:
```
The community has recognized the need ...
```
adding a transition like:
```
Specifically, the Stacks community has recognized the need ...
```

wileyj commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,532 @@
+Wallets implementing the new standard may maintain the previous system to support legacy applications during a transition period or indefinitely.
+Existing applications using the current Auth system should continue to operate, but immediate changes are recommended once this SIP is ratified.
+
+## Reference Implementations
```
if there is one, i think it would be helpful to share that information here rather than solely describing how the implementation should work. 
is there an existing implementation of this proposal?

i would argue that most of this section should be moved to the Specifications section, since that's what they appear to be defining: specifications, not implementation details.

jcnelson replied:
Yeah, let's link to a git tag in these projects so readers can have an example of a compliant artifact to study.

wileyj commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,532 @@
+- Addresses are serialized as Stacks c32-encoded strings.
+- Clarity values, post-conditions, and transactions are serialized to bytes (defined by SIP-005) and used as hex-encoded strings.
+
+### Methods
```
+1 - made a comment above about this exact thing before seeing you state the same here.

janniks replied:
The method definitions? They are sort of the largest part of the spec. But we can add more on specs for provider discovery and implementation

wileyj replied:
yes - i think it's more accurate to have them in the specifications section rather than the implementation section. 
the current text doesn't describe or define an implementation of the specification is my reasoning. 

i referred to https://github.com/stacksgov/sips/blob/main/sips/sip-000/sip-000-stacks-improvement-proposal-process.md?plain=1#L165-L170  for my reasoning here.

obycode replied:
It seems like a better ordering might be:

```diff
diff --git a/sips/sip-030/sip-030-wallet-interface.md b/sips/sip-030/sip-030-wallet-interface.md
index 7e0050a..d825dd5 100644
--- a/sips/sip-030/sip-030-wallet-interface.md
+++ b/sips/sip-030/sip-030-wallet-interface.md
@@ -49,16 +49,6 @@ The proposed changes are listed as follows:
 
 > **Note**: The methods can be used via the mentioned browser object (i.e., via the `window.WalletProvider.request` method), but also through similar interfaces like WalletConnect.
 
-## Backwards Compatibility
-
-The implementation of this proposal is not necessarily backward compatible.
-Notably, app-specific private keys for domains have been removed from the connect protocol, but can still be accessed via the `gaiaAppKey` field in `stx_getAccounts`.
-
-Wallets implementing the new standard may maintain the previous system to support legacy applications during a transition period or indefinitely.
-Existing applications using the current Auth system should continue to operate, but immediate changes are recommended once this SIP is ratified.
-
-## Reference Implementations
-
 ### Notes on Serialization
 
 To ensure serializability, consider these notes:
@@ -522,6 +512,16 @@ Listed below are some examples of the potentially unclear representations:
 Wallets can register their aliased provider objects however they see fit.
 For example, using the WBIP-004[^3] standard or Wallet Standard[^14].
 
+## Backwards Compatibility
+
+The implementation of this proposal is not necessarily backward compatible.
+Notably, app-specific private keys for domains have been removed from the connect protocol, but can still be accessed via the `gaiaAppKey` field in `stx_getAccounts`.
+
+Wallets implementing the new standard may maintain the previous system to support legacy applications during a transition period or indefinitely.
+Existing applications using the current Auth system should continue to operate, but immediate changes are recommended once this SIP is ratified.
+
+## Reference Implementations
+
 ## Activation
 
 This SIP is considered _Ratified_ after at least two wallet providers with significant user adoption (> 10,000 monthly active users) have implemented and launched the new standard.
```

wileyj commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,538 @@
+
+## Activation
+
+This SIP is considered _Ratified_ after Xverse and Leather (currently the largest wallets in the Stacks ecosystem) have implemented and launched the new standard.
```
I'm not a fan of this phrasing - since it gives the impression that these changes are meant to benefit the two largest wallet providers. 

can it be rephrased so something more agnostic, like:
```
This SIP is considered _Ratified_ after at least two wallet providers have implemented and launched the new standard.
```
i'm not sure if it's possible to add some parameters here like "2 wallet providers with x number of downloads/users/dau etc", but that would help qualify it better.

janniks commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,532 @@
+
+The proposed changes are listed as follows:
+
+Specify [JSON-RPC 2.0](https://www.jsonrpc.org/specification) compatible methods and payloads for wallet interaction.
```
Made it more clear now.

janniks committed:
chore: add more details on the scope of the specification

janniks committed:
chore: update network functions

janniks committed:
chore: mention clarity value type

janniks committed:
chore: add auth to related work

janniks committed:
chore: udpate activation phrasing

janniks committed:
chore: mention sbtc

friedger commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,571 @@
+- `txid`: `string` hex-encoded
+- `transaction`: `string` hex-encoded raw transaction
+
+#### Method `stx_transferSip9Nft`
```
stx_transferSip13Sft could be added as well.

janniks replied:
I can't tell if here SFTs should be treated as FTs

> Transfer
>
> (transfer ((token-id uint) (amount uint) (sender principal) (recipient principal)) (response bool uint))
> 
> Transfer a token from the sender to the recipient. It is recommended to leverage Clarity primitives like ft-transfer? to help safeguard users. The function should return (ok true) on success or an err response containing an unsigned integer on failure. The failure codes follow the existing conventions of stx-transfer? and ft-transfer?.

friedger replied:
SFTs have one more parameter, therefore, it should be specified as well.

Hero-Gamer replied:
@janniks any comment on this. Last comment to resolve before merging.

friedger commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,571 @@
+
+Errors thrown by request methods should match existing JSON-RPC 2.0 error codes.
+This way, the user or an intermediary library can handle them in a standardized way.
+Otherwise, no additional error codes are defined in this SIP.
```
Common errors like "user denied request" should have a defined error code
Leather uses 4001.

janniks replied:
Added

kyranjamie commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,571 @@
+
+SIP Number: `030`
+
+Title: Integration of a Modern Stacks Wallet Interface Standard
```
```suggestion
Title: Definition of a Modern Stacks Wallet Interface Standard
```

kyranjamie commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,571 @@
+- `txid`: `string` hex-encoded
+- `transaction`: `string` hex-encoded raw transaction
+
+#### Method `stx_signTransaction`
+
+Parameter properties
+
+- `transaction`: `string` hex-encoded raw transaction
```
```suggestion
#### Method `stx_signTransaction`

Parameter properties

- `transaction`: `string` hex-encoded raw transaction
- `broadcast?`: `boolean` whether transaction is to be broadcast, defaults to false
```

jcnelson commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,571 @@
+- Bytes are serialized as hex-encoded strings (without a 0x prefix).
+- Predefined formats from previous SIPs are used where applicable.
+- Addresses are serialized as Stacks c32-encoded strings.
+- Clarity values, post-conditions, and transactions are serialized to bytes (defined by SIP-005) and used as hex-encoded strings.
```
SIP-003 is also relevant here

jcnelson commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,571 @@
+- `fee?`: `number | string` BigInt constructor compatible value
+- `nonce?`: `number | string` BigInt constructor compatible value
+- `postConditions?`: `PostCondition[]`, defaults to `[]`
+  - where `PostCondition` is `string | object` hex-encoded or JSON representation
```
Is there further structure to `PostCondition` you can put here, or link to?

janniks replied:
Great point added a note that this and ClarityValue are defined later in the document.

jcnelson commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,571 @@
+
+- `recipient`: `string` address, Stacks c32-encoded
+- `asset`: `string` address, Stacks c32-encoded, with contract name suffix
+- `assetId`: `ClarityValue`
```
Is there a doc somewhere to what a `ClarityValue` type is?  If so, can you link to it in the `Parameter properties` section above?

jcnelson reviewed (commented):
I think this SIP is in a good place, with only minor comments remaining.

Can the technical CAB take a look and consider advancing this SIP? cc @obycode @kantai et al.

janniks committed:
chore: add note on type definitions

janniks committed:
chore: add broadcast

Co-authored-by: kyranjamie <kyranjamie@users.noreply.github.com>

janniks committed:
chore: add broadcast common note

Hero-Gamer commented:
"**_This SIP is considered Ratified after at least two wallet providers with significant user adoption (> 10,000 monthly active users) have implemented and launched the new standard._**"

Has this been met? Do we have 2 wallet provider confirm to implement? If so, maybe useful to drop their confirmation here? 

Anything else outstanding? If not I will ask the Technical CAB chair to table it for official vote.

kyranjamie commented:
Leather now supports SIP-30

janniks commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,538 @@
+- `signature`: `string` hex-encoded
+- `publicKey`: `string` hex-encoded
+
+#### Method `stx_getAddresses`
+
+> **Note**: This method can be used similarly to legacy "connect" methods, where the first account selection also acts as an approval of the site/domain.
+
+Result properties
+
+- `addresses`: `{}[]`
+  - `address`: `string` address, Stacks c32-encoded
+  - `publicKey`: `string` hex-encoded
```
👍

janniks committed:
docs: add feedback

janniks committed:
docs: add copyright

janniks committed:
docs: change title

friedger commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,611 @@
+- `asset`: `string` address, Stacks c32-encoded, with contract name suffix
+- `assetId`: `ClarityValue`
+
+`where`
```
```suggestion
where
```

janniks committed:
docs: update style

Co-authored-by: Friedger <friedger@gmail.com>

janniks committed:
docs: update

jiga commented on sips/sip-030/sip-030-wallet-interface.md:
```diff
@@ -0,0 +1,611 @@
+
+Type: Standard
+
+Status: Draft
```
```suggestion
Status: Recommended
```

Technical CAB has voted and approved this SIP. The Status is now Recommended for the Stearing committee review

jiga reviewed (changes requested):
@jcnelson @MarvinJanssen @GinaAbrams_

Hero-Gamer commented:
> Leather now supports SIP-30

Great it's passed CAB vote!

Last thing, @kyranjamie & @janniks:

To help the Steering Committee review whether the activation criteria have been met, it was mentioned that there are two wallet providers, with Leather being one of them. How can we verify who the second wallet provider is and confirm that they have implemented and launched the new standard?

_Activation
This SIP is considered Ratified after at least two wallet providers with significant user adoption (> 10,000 monthly active users) have implemented and launched the new standard._

jcnelson committed:
chore: SIP-030 is now recommended

Co-authored-by: _jiga <i.digg.tech@gmail.com>

314159265359879 commented:
I believe someone from Xverse can confirm in this topic that they have implemented/launched SIP030 support. @aryzing can you do that?

aryzing commented:
Thanks for the ping @314159265359879. Current SIP030 support in Xverse Wallet:

- [done] stx_transferStx
- [not implemented] stx_transferSip10Ft
- [not implemented] stx_transferSip9Nft
- [done] stx_callContract
  - `functionArgs` only support hex-encoded strings for arguments, not object clarity values.
- [done] stx_deployContract
- [done] stx_signTransaction
- [done] stx_signMessage
- [done] stx_signStructuredMessage
  - `message` only supports hex encoded strings, not objects
  - `domain` only supports strings
- [done] stx_getAddresses
- [incompatible] stx_getAccounts
  - Returning same payload under `addresses` key rather than `accounts` key.
- [not implemented] stx_getNetworks
- [not implemented] stx_updateProfile

Xverse Wallet doesn't support Stacks-only events, although we do have wallet-wide account- and network-change events.

314159265359879 commented:
Thanks Edu, that looks great. We have two wallets providers (Leather and Xverse) that have implemented and launched wallets with SIP030 support. 

I don't think leaving some specific parts of SIP030 for the future by specific wallets should prevent this SIP from moving to ratified. I believe support for this SIP has been sufficiently established for ratification.

friedger commented:
@314159265359879  I disagree. If the apis are not fully implemented then the wallet is not compatible. How can developers rely on the SIP if it only works partly.

Maybe the sip is too complex?

aryzing commented:
My understanding is that SIPs have a [workflow](https://tc39.es/process-document/) to accommodate progressive levels of finalization and adoption. Seeing this SIP is marked as "Recommended", it would make sense to merge it in its current form and when adoption increases, we can move it along the workflow as needed?

friedger commented:
[SIP-000](https://github.com/stacksgov/sips/blob/main/sips/sip-000/sip-000-stacks-improvement-proposal-process.md) is not clear about when proposals should be merged to main.

I think we can even set the status to `Activation-In-Progress`

Hero-Gamer commented:
**My SIP-000 interpretation and SIP Repo's current approach**
* Perhaps not explicitly mentioned, but the SIP-000 description implies that once all activation criteria are met, the Steering Committee can transition the status to 'Ratified.' `

* Being "Ratified" logically indicates that all checks and approvals have been completed, making the proposal eligible for **merging** into the main branch. That's how it's been generally handled historically, i.e. would only be merging once it's been ratified. 

Extract link: https://github.com/stacksgov/sips/blob/main/sips/sip-000/sip-000-stacks-improvement-proposal-process.md#overseeing-sip-activation-and-ratification
![image](https://github.com/user-attachments/assets/54e0e54f-64d1-4ebc-8cd5-8f2a119b8645)

**Possible future approach following Bitcoin's approach**
* Although it is also possible to take Bitcoin's [BIP repo](https://github.com/bitcoin/bips) approach, you can see many BIPs not "Final" status yet, however they are merged into main, under status such as draft, proposed. 

* A BIP being merged means it’s recognized and part of the record — not that it’s final or implemented. Only after consensus, deployment, and sometimes community/user signaling, does a BIP become Final.

**My 2 Sats on how to treat merging for SIP-030**
* Seems no harm to **merge** under "Activation-In-Progress" status to make it more publicly discoverable, because it seems useful to have SIP-030 be easily accessible in SIP Repo's main list.

* Think of it like a **bulletin board**, it’s now public, but not yet "**law**". Available for review, discussion, tracking, progressive finalization of specification for future ratification.

* Nevertheless setting status to `Activation-In-Progress` definitely still makes sense if you think the criteria has not been "unambiguously met" yet.

Perhaps one of the Steering Committee or Technical CAB members can weight in here.

wileyj commented:
> **My SIP-000 interpretation and SIP Repo's current approach**
> 
> * Perhaps not explicitly mentioned, but the SIP-000 description implies that once all activation criteria are met, the Steering Committee can transition the status to 'Ratified.' `
> * Being "Ratified" logically indicates that all checks and approvals have been completed, making the proposal eligible for **merging** into the main branch. That's how it's been generally handled historically, i.e. would only be merging once it's been ratified.
> 
> Extract link: https://github.com/stacksgov/sips/blob/main/sips/sip-000/sip-000-stacks-improvement-proposal-process.md#overseeing-sip-activation-and-ratification ![image](https://private-user-images.githubusercontent.com/92864860/439879400-54e0e54f-64d1-4ebc-8cd5-8f2a119b8645.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDYyMDE0NDUsIm5iZiI6MTc0NjIwMTE0NSwicGF0aCI6Ii85Mjg2NDg2MC80Mzk4Nzk0MDAtNTRlMGU1NGYtNjRkMS00ZWJjLThjZDUtOGYyYTExOWI4NjQ1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTAyVDE1NTIyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc0ZjY2ZGRkNGY1ODUwOTZlMmYxNTE2ODNkMjJjNTYwMjE5YjMyYTk2NGJkNTJjYmMzMzEwN2ZmNzYwZGNiNjImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Hgiuid4RDOUt_mFHFLopAABbodi04LkjEymdUzrqiXU)
> 
> **Possible future approach following Bitcoin's approach**
> 
> * Although it is also possible to take Bitcoin's [BIP repo](https://github.com/bitcoin/bips) approach, you can see many BIPs not "Final" status yet, however they are merged into main, under status such as draft, proposed.
> * A BIP being merged means it’s recognized and part of the record — not that it’s final or implemented. Only after consensus, deployment, and sometimes community/user signaling, does a BIP become Final.
> 
> **My 2 Sats on how to treat merging for SIP-030**
> 
> * Seems no harm to **merge** under "Activation-In-Progress" status to make it more publicly discoverable, because it seems useful to have SIP-030 be easily accessible in SIP Repo's main list.
> * Think of it like a **bulletin board**, it’s now public, but not yet "**law**". Available for review, discussion, tracking, progressive finalization of specification for future ratification.
> * Nevertheless setting status to `Activation-In-Progress` definitely still makes sense if you think the criteria has not been "unambiguously met" yet.
> 
> Perhaps one of the Steering Committee or Technical CAB members can weight in here.

I tend to agree, but i would defer to the steering committee on if this should be the process for sips like this. 
i'll see if i can mention this in a few channels, but outside of this specific point i think we're in a mergeable state, correct (minus some missing PR approvals)?

MarvinJanssen commented:
As a (occasionally observing) member of the steering committee, I agree that it makes sense to merge non-ratified SIPs to the main branch once they reach a certain phase in the workflow. I would agree with @obycode's suggestion to merge `Recommend` and up if not for one issue. SIPs in the `Recommended` phase would be at risk of staying in that state for an indefinite time. The step from `Recommended` to `Activation-In-Progress` allows the steering committee to guard the SIP process and make sure that the SIP adheres to SIP000. If a Recommended SIP is merged then it is unclear how the steering committee would provide feedback and approval to move it to the next phase. Perhaps the authors never open a follow-up PR to advance the SIP. In that case, the onus would now be on the steering committee members (or who?) to open a PR to change the SIP to `Rejected` or `Withdrawn`. How long would we wait before making the determination? 

Additionally, it brings with it a bit more work to keep things clean. The README should be updated with the SIP list replaced with a table so that we can indicate the status of each SIP.

Hero-Gamer commented:
Thanks all! Looks like we have a rough consensus already!

I suggest we move forward with merging SIP-030 in which is "Activation-In-Progress" already as it's [voted & approved by the Tech CAB back in Feb/Mar](https://github.com/stacksgov/sips/blob/main/considerations/minutes/technical-cab/2025-03-05-sip-030.md), and _if any future SIPs are at "Recommended" stage that are_ **merge worthy**, we could review the guideline again as and when. 
For the moment let's run with only merging "Activation-In-Progress" as Marvin suggested.

Hero-Gamer commented:
Tech CAB reviewed this. Minutes doc here: https://github.com/stacksgov/sips/blob/main/considerations/minutes/technical-cab/2025-03-05-sip-030.md

I think it's good to merge into repo and placed under in the process of being activated? if there isn't anything missing?

jcnelson commented:
Circling back to this, have two or more wallets with over 10,000 users added support for this SIP?

314159265359879 reviewed (approved):
Lets merge this with "Activation in progress" label @brice-stacks, I agree with [Brice](https://github.com/stacksgov/sips/pull/166#issuecomment-2847720160), [Marvin](https://github.com/stacksgov/sips/pull/166#issuecomment-2852208919) en [Friedger](https://github.com/stacksgov/sips/pull/166#issuecomment-2846765895)  that is a good idea. we can change the status once Xverse has 100% compatibility with the SIP.

Once it is in the main branche it will also be easier to make the [proposed change](https://github.com/stacksgov/sips/pull/166#issuecomment-4558778378) for the new post-condition mode. @janniks Would you mind checking my comment see if that makes sense and perhaps create the necessary change PR after this is merged?
